File: p005m/longest_palindrome.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function (s) {
  let findPalindrome = function (l, r) {
    while (l >= 0 && r < s.length && s[l] === s[r]) {
      l--;
      r++;
    }
    return [l + 1, r - l - 1]; // [start, length]
  };
  let [maxstart, maxlen] = [0, 0];
  for (let i = 0; i < s.length; ++i) {
    let [oddstart, oddlen] = findPalindrome(i, i);
    let [evenstart, evenlen] = findPalindrome(i, i + 1);
    if (oddlen > maxlen) [maxstart, maxlen] = [oddstart, oddlen];
    if (evenlen > maxlen) [maxstart, maxlen] = [evenstart, evenlen];
  }
  return s.substr(maxstart, maxlen);
};
// TEST
[
  ['babad', 'bab'],
  ['cbbd', 'bb'],
  ['a', 'a'],
  ['ac', 'a'],
].forEach(([s, expected]) => {
  const actual = longestPalindrome(s);
  console.log('Longest palindrome of "' + s + '" ->', actual);
  console.assert(actual === expected);
});

File: p005m/longest_palindrome.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        max_start, max_len = n - 1, 1
        dp = [[False for _ in range(n)] for _ in range(n)]
        for i in range(n - 1, -1, -1):
            for j in range(i, n):
                dp[i][j] = s[i] == s[j] and (j - i < 3 or dp[i + 1][j - 1])
                if dp[i][j] and j - i + 1 > max_len:
                    max_start, max_len = i, j - i + 1
        return s[max_start : max_start + max_len]


# TESTS
for s, expected in [("babad", "aba"), ("cbbd", "bb"), ("a", "a"), ("ac", "c")]:
    sol = Solution()
    actual = sol.longestPalindrome(s)
    print(f"Longest palindrome of '{s}' -> {actual}")
    assert actual == expected
